File: 백준/Silver/2512. 예산/예산.py
import sys
input = sys.stdin.readline

n = int(input())
city_list = list(map(int, input().split()))
total_m = int(input())

def get_sum_m(k):
    sum_money = 0
    for city in city_list:
        if city < k:
            sum_money += city
        else:
            sum_money += k
    return sum_money

def get_max_m():
    left = 0
    right = total_m
    answer = total_m
    while left <= right:
        mid = (left + right) // 2
        sum_m = get_sum_m(mid)
        
        if sum_m < total_m:
            answer = mid
            left = mid + 1
        elif sum_m > total_m:
            right = mid - 1
        else:
            return mid
    
    return answer

if get_sum_m(total_m) <= total_m:
    print(max(city_list))
else:
    print(get_max_m())

File: 백준/Silver/2512. 예산/예산.js
const input = require('fs').readFileSync(process.platform === 'linux' ? '/dev/stdin' : __dirname + '/input.txt').toString().split('\n')

const N = Number(input[0])
const request_budgets = input[1].split(' ').map(Number)
request_budgets.sort((a, b) => a - b)
const max_budget = Number(input[2])

let answer = 0
let left = 0
let right = Math.max(...request_budgets)
while(left <= right) {
  const mid = Math.floor((left + right) / 2)
  const bugets = divide_budget(mid)
  
  if (bugets <= max_budget) {
    answer = Math.max(answer, mid)
    left = mid + 1
  } else {
    right = mid - 1
  }
}

console.log(answer)


function divide_budget(limit) {
  let sum = 0
  for (const buget of request_budgets) {
    sum += Math.min(limit, buget)
  }

  return sum
}